### ISSUE TITLE
Proposal: Framework-level support for saving/restoring UI state

### ISSUE THREAD

**Issue by micahl** [microsoft/microsoft-ui-xaml#1061]
# Proposal: Framework-level support for saving/restoring UI state

## Summary

In applications that use a navigation model where users move between pages users expect to be able to navigate back to a previous page and have it appear as it did before (perhaps with refreshed content). 

## Rationale

An easy option might be to cache the page, however, this consumes resources and may not be an option for some apps.  An alternative that may be required is for the app to save and restore the state of the UI (e.g. scroll offset, layout, focus, selection).  In basic scenarios this may be simple to do.  It's more complex when dealing with one or more virtualizing layouts.  XAML currently provides helpers such as the [ListViewPersistenceHelper ](https://docs.microsoft.com/en-us/uwp/api/Windows.UI.Xaml.Controls.ListViewPersistenceHelper) to make this easier when using a ListView/GridView.  No such helper exists when using ItemsRepeater.  Rather than another helper, support for saving/restoring the UI state should be provided by the framework.  

## Scope

| Capability | Priority |
| :---------- | :------- |
| Developers can easily save/restore the UI state (e.g. on back navigation or app reactivation) | Must |
| Developers control when the control state is updated | Must |
| PasswordBox must not leak any sensitive information | Must |
| A subset of controls should leverage this framework-level support (e.g. ScrollViewer, ItemsRepeater, TextBox). | Should |

## Important Notes

* Controls that need to store/retrieve their state would have to be in the live visual tree.
* Sometimes the data to present on a page will have changed between page navigation (e.g. an item was deleted or inserted).  Restoring a scroll position may need to be relative to the position of an item in a virtualized list and have that position be maintained as additional items are realized and repositioned (i.e. items that contain images may change size on subsequent ticks since the images are decoded asynchronously and appear a few ticks later).

#### Default Control Properties to be Stored
| Control | Properties | Publicly accessible |
| :------- | :---------- | :----------------- |
| TextBox, RichEditBox | Content, Caret position, and Selection | Yes |
| ScrollViewer | Scroll position and Anchor element | Yes |
| ItemsRepeater | Anchor element | Yes |
| All controls | Focus | Yes |

## Open Questions

* Are there 3rd party controls that would benefit from a general mechanism to participate in saving/restoring their internal state?
* Is fine-grained control over which properties/state to include needed?  For example, restoring a TextBox's content but not its selection or caret position?

**Comment by mdtauk** [comment]
Anything that can make it easy and automatic would be welcome I imagine, with a way to opt out.

Ideally it would be achievable in XAML, perhaps on the **Page** element, some property like **ManageState** set to True by default in a new WinUI 3.0 project

**Comment by micahl** [comment]
@mdtauk, agreed on the importance of making it easy.  In terms of prior art, UIKit has [this](https://developer.apple.com/documentation/uikit/view_controllers/preserving_your_app_s_ui_across_launches).  SwiftUI obviously uses a different mechanism and is designed around the notion of state being lifted out of the "controls".

Removing myself as the assignee.  This proposal will have to sit on the backlog for now as its dependent on WinUI 3.0.

**Comment by harvinders** [comment]
@micahl  Can we also consider nested restoration of UI state as described in the use case [here](https://github.com/microsoft/microsoft-ui-xaml/issues/1848#issuecomment-574940193)

**Comment by micahl** [comment]
@harvinders, it was a scenario in mind when I authored the proposal.  As I'm no longer directly involved in the project I'll defer to the WinUI team on where this proposal goes.

**Comment by ranjeshj** [comment]
@anawishnoff as FYI. @harvinders using the Anchor and bring into view would be the way to do it for ScrollViewer/ItemsRepeater. This is a proposal to make it simpler especially in nested cases and possibly provide some APIs similar to persistence helper.

**Comment by dotMorten** [comment]
> Are there 3rd party controls that would benefit from a general mechanism to participate in saving/restoring their internal state?

Yes! I have a map control. When I navigate away from the page and back, I'd like to make sure I restore the location where the user had zoomed to, which map was open, which layers were turned on, if GPS location was turned on etc.
The map is however extremely expensive to keep in view, so I'd prefer not having to cache.
